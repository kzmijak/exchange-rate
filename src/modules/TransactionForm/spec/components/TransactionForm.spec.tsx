import { describe, it, expect, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import {
  TransactionForm,
  TransactionFormProps,
} from "../../src/components/TransactionForm";
import userEvent from "@testing-library/user-event";

const OK_TITLE = "proper title";
const OK_AMOUNT = 500;

describe("TransactionForm", () => {
  const titleInputDOM = () => screen.getByRole("textbox", { name: /Title/ });
  const amountInputDOM = () => screen.getByRole("textbox", { name: /Amount/ });
  const submitButtonDOM = () => screen.getByRole("button", { name: /Add/ });
  const submitHandler = vi.fn();

  const Component = (props: Partial<TransactionFormProps>) => (
    <TransactionForm onSubmit={submitHandler} {...props} />
  );

  afterEach(() => {
    submitHandler.mockReset();
  });

  it("should match snapshot", () => {
    expect(render(<Component />).container).toMatchSnapshot();
  });

  it("should call onSubmit when fills form is filled properly", async () => {
    render(<Component />);
    await userEvent.type(titleInputDOM(), OK_TITLE);
    await userEvent.type(amountInputDOM(), OK_AMOUNT.toString());
    await userEvent.click(submitButtonDOM());
    expect(submitHandler).toHaveBeenCalledOnce();
  });

  it("should properly forward focus with tabs", async () => {
    render(<Component />);
    await userEvent.click(titleInputDOM());
    expect(titleInputDOM()).toHaveFocus();
    await userEvent.tab();
    expect(amountInputDOM()).toHaveFocus();
    await userEvent.tab();
    expect(submitButtonDOM()).toHaveFocus();
  });

  it("should trigger submit on enter click", async () => {
    render(
      <Component defaultValues={{ amount: OK_AMOUNT, title: OK_TITLE }} />
    );
    await userEvent.click(titleInputDOM());
    await userEvent.keyboard("{enter}");
    expect(submitHandler).toHaveBeenCalledOnce();
  });

  describe("title validation", () => {
    it("is required", async () => {
      render(<Component defaultValues={{ amount: OK_AMOUNT }} />);
      await userEvent.click(submitButtonDOM());
      expect(screen.getByText("Title is required")).toBeVisible();
      expect(submitHandler).not.toHaveBeenCalled();
    });

    it("has to be at least 5 characters long", async () => {
      render(
        <Component defaultValues={{ amount: OK_AMOUNT, title: "1234" }} />
      );
      await userEvent.click(submitButtonDOM());
      expect(
        screen.getByText("Title should have at least 5 characters.")
      ).toBeVisible();
      expect(submitHandler).not.toHaveBeenCalled();
      await userEvent.type(titleInputDOM(), "5");
      expect(
        screen.queryByText("Title should have at least 5 characters.")
      ).not.toBeInTheDocument();
      await userEvent.click(submitButtonDOM());
      expect(submitHandler).toHaveBeenCalledOnce();
    });
  });

  describe("amount validation", () => {
    it("is required", async () => {
      render(<Component defaultValues={{ title: OK_TITLE }} />);
      await userEvent.click(submitButtonDOM());
      expect(screen.getByText("Amount is required")).toBeVisible();
      expect(submitHandler).not.toHaveBeenCalled();
    });
  });
});
