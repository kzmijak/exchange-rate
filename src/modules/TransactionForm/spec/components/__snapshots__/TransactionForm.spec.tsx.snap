// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TransactionForm > should match snapshot 1`] = `
<div>
  <form
    class="MuiStack-root css-3tz32n-JoyStack-root"
  >
    <div
      class="MuiStack-root css-1tb7emz-JoyStack-root"
    >
      <div
        class="MuiFormControl-root MuiFormControl-vertical MuiFormControl-sizeMd MuiStack-root css-9cy8of-JoyFormControl-root-JoyStack-root"
      >
        <label
          class="MuiFormLabel-root css-1ftzm8l-JoyFormLabel-root"
          for=":r0:"
          id=":r0:-label"
        >
          Title of transaction
        </label>
        <div
          class="MuiBox-root css-1rr4qq7"
        >
          <div
            class="MuiInput-root MuiInput-fullWidth MuiInput-variantOutlined MuiInput-colorNeutral MuiInput-sizeMd MuiInput-formControl css-gsnhoa-JoyInput-root"
          >
            <input
              class="MuiInput-input css-1gw9vc6-JoyInput-input"
              id=":r0:"
              name="title"
              placeholder="ex. Busk ticket"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiFormControl-vertical MuiFormControl-sizeMd MuiStack-root css-9cy8of-JoyFormControl-root-JoyStack-root"
      >
        <label
          class="MuiFormLabel-root css-1ftzm8l-JoyFormLabel-root"
          for=":r1:"
          id=":r1:-label"
        >
          Amount (in PLN)
        </label>
        <div
          class="MuiBox-root css-1rr4qq7"
        >
          <div
            class="MuiInput-root MuiInput-variantOutlined MuiInput-colorNeutral MuiInput-sizeMd MuiInput-formControl css-1vip6bn-JoyInput-root"
            inputmode="numeric"
          >
            <input
              class="MuiInput-input css-1moyr10-JoyInput-input"
              id=":r1:"
              name="amount"
              placeholder="ex. 25.99"
              type="text"
              value="0"
            />
            <div
              class="MuiInput-endDecorator css-x3cgwv-JoyInput-endDecorator"
            >
              z≈Ç
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiStack-root css-jkms5x-JoyStack-root"
    >
      <button
        class="MuiButton-root MuiButton-variantSolid MuiButton-colorPrimary MuiButton-sizeMd css-18m3si8-JoyButton-root"
        type="submit"
      >
        Add
      </button>
    </div>
  </form>
</div>
`;
